### ISSUE TITLE
Attribute not rendered by Marko is not preserved if component first rendered on the server

### ISSUE THREAD

**Issue by maximilianschmitt** [marko-js/marko#644]
Hi again!

For a project I'm looking at building with Marko 4 (I'm on `Marko 4.2.0`), I want an `input` element to be autosized so that the width of the input always matches the width of its value. The following works fine for the most part:

```
static var autosize = require('autosize-input')

class {
  onMount() {
    // resize the input on input
    autosize(this.getEl('input'))
  }

  onChange(e) {
    this.emit('change', e.target.value)
  }
}

div
  input key='input' value=input.value on-input('onChange')
  -- ${input.value}
```

However, on the first input (e.g. entering a space) the `style` attribute (that the autosize-module sets) seems to be removed:

![marko-autosize](https://cloud.githubusercontent.com/assets/3694962/24577712/37153aec-16d3-11e7-971f-91822d449d6b.gif)

Full code: https://github.com/maximilianschmitt/lasso-test

I wonder if this is intentional behaviour? I've also tried manually resizing in the `onRender` and `onUpdate` hooks but this causes flickering.

Now finally I modified `autosize-input` to return the style and use Marko's state to update the `style` together with the `input`'s value. This solves my issue above but with this pattern the cursor of the `input` jumps to the end if I enter a character:

```
static var autosize = require('autosize-input')

class {
  onCreate() {
    this.state = { autosize: {} }
  }

  onMount() {
    var component = this

    // on every input, call my callback with the calculated style
    autosize(this.getEl('input'), null, function (style) {
      component.state.autosize = style
      component.setStateDirty('autosize')
    })
  }

  onChange(e) {
    this.emit('change', e.target.value)
  }
}

div
  input key='input' style=state.autosize value=input.value on-input('onChange')
```

![marko-autosize2](https://cloud.githubusercontent.com/assets/3694962/24577803/4c669560-16d5-11e7-9c4d-ed088b85a8d6.gif)

Full code: https://github.com/maximilianschmitt/lasso-test/tree/cursor-jump

Thanks for your help! :)

**Comment by patrick-steele-idem** [comment]
Thank you for the detailed problem report and the working app. Very helpful!

I put in a fix to make this issue less common, but it has to do with the UI component first being rendered on the server and Marko not knowing that an attribute was added by an external library/code. When rendered initially in the browser this would not have be an issue. The fix skips removing non-Marko rendered attributes for keyed elements. The `key` is required to know for sure that the two elements being diffed are the matching elements.

New version published with fix: `marko@4.2.1`

**Comment by maximilianschmitt** [comment]
Hey, thanks! Interesting to hear what caused the issue. `4.2.1` fixed it for me. :) Thank you!
